src/Microsoft.AspNetCore.Server.Kestrel/BadHttpResponseException.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Http;
+
+namespace Microsoft.AspNetCore.Server.Kestrel
+{
+    public sealed class BadHttpResponseException : InvalidOperationException
+    {
+        private BadHttpResponseException(string message)
+            : base(message)
+        {
+
+        }
+
+        internal static BadHttpResponseException GetException(ResponseRejectionReasons reason)
+        {
+            BadHttpResponseException ex;
+            switch (reason)
+            {
+                case ResponseRejectionReasons.HeadersReadonlyResponseStarted:
+                    ex = new BadHttpResponseException("Headers are read-only, response has already started.");
+                    break;
+                default:
+                    ex = new BadHttpResponseException("Bad response.");
+                    break;
+            }
+
+            return ex;
+        }
+
+        internal static BadHttpResponseException GetException(ResponseRejectionReasons reason, string value)
```

Comment:
You could make `value` a generic arg here, so you wouldn't have to call `ToString()` on `StatusCode` below.